# Tarea 1 Bases de Datos 2023-1 Grupo XX

_Recuerde modificar esta celda y la siguiente! (Borre esta línea también)._

## Integrantes:
* Nombre Integrante 1 - Rol-1
* Nombre Integrante 2 - Rol-2

## Requisitos para Correr el Notebook

Para poder utilizar este Notebook se debe tener instalado:
- **Python 3**
- **Jupyter**
- La librería **ipython-sql** (Poder ejecutar sql en Jupyter Notebook)
- PostgreSQL y su conector desde Python(**psycopg2**)

**IMPORTANTE:** Este Jupyter Notebook se conectará a su servicio de Base de Datos PostgreSQL de manera **local**, por lo que **no funcionará correctamente en el entorno Google Collab.**

Para instalar las dependencias, ejecute (**recuerde actualizar pip3**):

In [6]:
! pip3 install ipython-sql
! pip3 install psycopg2

#Use to generate names
! pip3 install names

## Importar dependencias

In [7]:
from datetime import datetime
import random as r
import names

#Librerías para PostgreSQL
import psycopg2

Para conectarse a la base de datos, completen los datos relevantes en la variable DATABASE_URL.

* Cambie `user` por el usuario que utilizó en Postgres para acceder a la Base de Datos. Por defecto `postgres`.
* Cambie `password` por la contraseña de dicho usuario. **Recuerde borrarla antes de entregar su tarea.**
* Cambie `port` por el puerto en que se ejecuta su servicio de Base de Datos. Por defecto `5432`.
* Cambie `database` por el nombre de la BD (Schema) que creó en Postgres.

Ejemplo: `postgresql://postgres:pass_segura@localhost:5432/tarea-bd`

In [8]:
%reload_ext sql
%env DATABASE_URL=postgresql://postgres:benjamin@localhost:5432/tareabase

env: DATABASE_URL=postgresql://postgres:benjamin@localhost:5432/tareabase


Para conectarse de forma declarativa a la Base de Datos, utilice el siguiente código:

In [9]:
#Recuerde cambiar los valores de USER, DBNAME y PASSWORD por los mismos de la celda superior.
connection = psycopg2.connect(
   host="localhost",
   user="postgres",
   dbname='tareabase',
   password="benjamin"
)

print(connection)
cursor = connection.cursor()

<connection object at 0x0000028D0350EF00; dsn: 'user=postgres password=xxx dbname=tareabase host=localhost', closed: 0>


## Crear tablas (36 pts.)

Aquí inserta el código SQL que creara las tablas de la base de datos:

### Tabla "facciones" (6 pts)

In [10]:
%%sql
CREATE TABLE facciones(

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    nombre VARCHAR(45) NOT NULL,

    descripcion VARCHAR(45) NOT NULL,
    
    PRIMARY KEY (id)

);
    

Done.


[]

### Tabla "trabajos" (6 pts)

In [11]:
%%sql
CREATE TABLE trabajos(
    
    id INTEGER GENERATED ALWAYS AS IDENTITY,
    
    nombre VARCHAR(45) NOT NULL,
    
    descripcion VARCHAR(45) NOT NULL,

    sueldo INTEGER NOT NULL,
    
    PRIMARY KEY (id)

);


 * postgresql://postgres:***@localhost:5432/tareabase
Done.


[]

### Tabla "personas" (6 pts)

In [12]:
%%sql
CREATE TABLE personas(

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    nombre VARCHAR(45) NOT NULL,

    apellidos VARCHAR(45) NOT NULL,

    fecha_nacimiento DATE NOT NULL,

    infectado BOOLEAN NOT NULL,

    conyugue INTEGER,

    PRIMARY KEY (id) ,
    
    FOREIGN KEY (conyugue) REFERENCES personas
);


 * postgresql://postgres:***@localhost:5432/tareabase
Done.


[]

### Tabla "cuentas_bancarias" (6 pts)

In [13]:
%%sql

CREATE TABLE cuentas_bancarias (

    id INTEGER GENERATED ALWAYS AS IDENTITY,

    tipo_de_cuenta VARCHAR(45) NOT NULL,

    banco VARCHAR(45) NOT NULL,

    saldo INTEGER NOT NULL,

    PRIMARY KEY (id),

    id_persona INTEGER, 

    FOREIGN KEY(id_persona) REFERENCES personas(id)

);

 * postgresql://postgres:***@localhost:5432/tareabase
Done.


[]

### Tabla "persona_pertenece_faccion" (6 pts)

In [14]:
%%sql
CREATE TABLE persona_pertenece_faccion(

    id_persona INTEGER,

    id_faccion INTEGER,

    PRIMARY KEY (id_persona, id_faccion),

    FOREIGN KEY (id_persona) REFERENCES personas (id),

    FOREIGN KEY (id_faccion) REFERENCES facciones (id)
    

);

 * postgresql://postgres:***@localhost:5432/tareabase
Done.


[]

### Tabla "persona_tiene_trabajo" (6 pts)

In [15]:
%%sql
CREATE TABLE persona_tiene_trabajo (
 
 id_persona INTEGER,

 id_trabajo INTEGER,
 
 estado BOOLEAN NOT NULL,

 ultima_vez_realizado TIMESTAMP NOT NULL,   
 
 FOREIGN KEY (id_persona) REFERENCES personas (id),
 
 FOREIGN KEY (id_trabajo) REFERENCES trabajos (id),
 
 PRIMARY KEY (id_persona , id_trabajo)

);


 * postgresql://postgres:***@localhost:5432/tareabase
Done.


[]

## Código Generador de Datos

Ejecute este código para cargar datos a las tablas creadas.

PD: Si el codigo tira error por "violar la unicidad de las PK" es por una coincidencia de numeros random, solamente volver a ejecutar el codigo

In [16]:
#Recuerde haber ejecutado todas las celdas anteriores!

# Generador de timestamp aleatorio.
def generate_timestamp_nacimiento():
    year = r.randint(1970,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

def generate_timestamp_trabajo():
    year = r.randint(2019,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

print("Borrando Datos Antiguos...")

# Limpia la base de datos.
connection.commit() # En caso de algun error
query = 'TRUNCATE TABLE "facciones", "trabajos", "personas", "cuentas_bancarias", "persona_pertenece_faccion", "persona_tiene_trabajo" RESTART IDENTITY'
cursor.execute(query)
connection.commit()

print("Generando Datos...")

facciones = [("FEDRA", "La Agencia Federal de Respuesta a Desastres"), ("Luciérnagas", "Grupo de milicias revolucionario"), ("Carroñeros", "Sin identificación"), ("Pythonia", "Grupo defensor de la libertad"), ("MaCdonia", "Los exiliados de la carceles"), ("JavaScriptia", "Grupo defensor de la paz")]
trabajos = [("Limpieza", "Encargado de limpiar las calles"),
("Limpiador de cloacas","Encargado de limpiar las cloacas"),
("Militar","Encargado de proteger la ciudad"),
("Explorador","Encargado de explorar la ciudad"),
("Médico","encargado de curar a los enfermos"),
("Profesor","Encargado de enseñar a los niños"),
("Cremador","Encargado de quemar los cadáveres"),
("Científico","Encargado de investigar"),
("Cocinero","Encargado de cocinar"),
("Carpintero","Encargado de construir"),
("Cerrajero","Encargado de reparar las cerraduras"),
("Cartero","Encargado de repartir los correos"),
("Cajero","Encargado de cobrar los impuestos")]
tipos_de_cuenta_bancaria = ["Cuenta Vista", "Cuenta Ahorro", "Cuenta Corriente", "Cuenta RUT", "Cuenta de Inversion", "Cuenta PrePago", "Cuenta de Credito"]
bancos = ["FedraBank S.A", "Banco Pedrito Pascal SpA", "Banco Pythonia S.A", "Banco de la Nación S.A", "Banco Santander SpA", "Banco BCI SpA", "Banco Estado SpA", "Banco Consorcio S.A", "Banco BBVA", "Banco Scotiabank"]

# Variables.
n_facciones = len(facciones)
n_trabajos = len(trabajos)
n_personas = 1000
n_cuentas_bancarias = 120
n_persona_pertenece_faccion = 300
n_persona_tiene_trabajo = 300
n_casados = 100

# Genera personas.
personas = []
for _ in range(n_personas):
    nombre = names.get_first_name()
    apellidos = names.get_last_name()+" "+names.get_last_name()
    fecha_nacimiento = generate_timestamp_nacimiento()
    infectado = r.choice([True, False])
    
    personas.append((nombre, apellidos, fecha_nacimiento, infectado, None))


# Genera cuentas bancarias.
cuentas_bancarias = []
for _ in range(n_cuentas_bancarias):
    tipo_de_cuenta = r.choice(tipos_de_cuenta_bancaria)
    banco = r.choice(bancos)
    saldo = r.randint(0, 10000)
    id_persona = r.choice(range(1, n_personas+1))
    
    cuentas_bancarias.append((tipo_de_cuenta, banco, saldo, id_persona))
    
    
# Generate persona_pertenece_faccion.
persona_pertenece_faccion = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_facciones+1))) for _ in range(1, n_persona_pertenece_faccion+1)])

# Generate persona_tiene_trabajo.
persona_tiene_trabajo_prev = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_trabajos+1))) for _ in range(1, n_persona_tiene_trabajo+1)])
persona_tiene_trabajo = [(id_persona, id_trabajo, r.choice([True, False]), generate_timestamp_trabajo()) for id_persona, id_trabajo in persona_tiene_trabajo_prev]
    
# SQL Statements.
inserts = [
    'INSERT INTO "facciones" ("nombre", "descripcion") VALUES (%s, %s);',
    'INSERT INTO "trabajos" ("nombre", "descripcion", "sueldo") VALUES (%s, %s, %s);',
    'INSERT INTO "personas" ("nombre", "apellidos", "fecha_nacimiento", "infectado", "conyugue") VALUES (%s, %s, %s, %s, %s);',
    'INSERT INTO "cuentas_bancarias" ("tipo_de_cuenta", "banco", "saldo", "id_persona") VALUES (%s, %s, %s, %s);',
    'INSERT INTO "persona_pertenece_faccion" ("id_persona", "id_faccion") VALUES (%s, %s);',
    'INSERT INTO "persona_tiene_trabajo" ("id_persona", "id_trabajo", "estado", "ultima_vez_realizado") VALUES (%s, %s, %s, %s);'
]

for faccion in facciones:
    cursor.execute(inserts[0], faccion)

for trabajo in trabajos:
    t = (trabajo[0], trabajo[1], r.randint(100, 5000))
    cursor.execute(inserts[1], t)

# Crear una lista con personas casadas y otra con personas solteras. donde se cambie la ide de la persona que se casó.

for persona in personas:
    cursor.execute(inserts[2], persona)
    
for cb in cuentas_bancarias:
    cursor.execute(inserts[3], cb)

for ppf in persona_pertenece_faccion:
    cursor.execute(inserts[4], ppf)
    
for ptt in persona_tiene_trabajo:
    cursor.execute(inserts[5], ptt)
    
connection.commit()


# Generar Matrimonios
n_repetidos = []

for i in range(n_casados):
    random = r.randint(1, n_personas)
    if random in n_repetidos:
        continue
    else:
        random2 = r.randint(1, n_personas)
        if random2 in n_repetidos:
            continue
        else:
            n_repetidos.append(random)
            n_repetidos.append(random2)
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random2, random))
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random, random2))
            connection.commit()

print("Todo Listo!")

Borrando Datos Antiguos...
Generando Datos...
Todo Listo!


## Consultas (64 pts.)

Escriba aquí las consultas pedidas. Recuerde agregar una descripción _breve_ de cómo funciona la solución. **Escriba las consultas de forma ordenada.**

**NOTA:** Cuando en una consulta diga "X cosa" significa que ustedes deben de elegir ese X por lo que para cualquier otro X debe de funcionar de igual forma.

### Consulta 1 (6 pts)
**ID, nombre y primer apellido de los habitantes de FEDRA**

Ejemplo:

|ID |  Nombre      |  Apellido  |
|---|-----------|-----------|
| 12 |Joel | Miller |
|21 | Ellie | Williams |
| ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._


In [ ]:
%%sql

SELECT
    personas.id , personas.nombre , split_part(apellidos, ' ', 1) AS primer_apellido
FROM 
    personas

INNER JOIN 
    persona_pertenece_faccion 
ON 
    personas.id = persona_pertenece_faccion.id_persona

INNER JOIN
    facciones
ON
    persona_pertenece_faccion.id_faccion = facciones.id
    

WHERE facciones.nombre = 'FEDRA'


### Consulta 2 (6 pts)
**Cantidad de personas nacidas después del inicio de la pandemia.**(inicio de la pandemia 01-12-2019)

Ejemplo:


| Cantidad     | 
|--------------|
| 292 | 

**Solución:** _Escriba aquí la descripción de su solución._

In [33]:
%%sql


SELECT 
    COUNT(*) AS cantidad 

FROM 
    personas 

WHERE 
    fecha_nacimiento > '2019-12-01';

 * postgresql://postgres:***@localhost:5432/tareabase
1 rows affected.


[(58,)]

### Consulta 3 (6 pts)
**ID, nombre, apellidos y saldo total de todas las personas**( Limitar a 20 filas para visualizar)

Ejemplo:

| ID    | Nombre | Apellido | Dinero | 
|----------------|--------|-|-|
| 13 | Tommy | Miller | 12501 |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [36]:
%%sql


SELECT
    id_persona , nombre , apellidos , saldo AS Dinero
FROM 
    personas 

INNER JOIN 
    cuentas_bancarias

ON  
    cuentas_bancarias.id_persona = personas.id

LIMIT 20;




 * postgresql://postgres:***@localhost:5432/tareabase
20 rows affected.


[(34, 'Mattie', 'Lydick Lee', 2526),
 (41, 'Virginia', 'Treat Lutz', 694),
 (51, 'Hazel', 'Shepherd Ball', 1669),
 (62, 'Dianne', 'Riley Cox', 3980),
 (65, 'Helen', 'Avila Johnson', 9603),
 (79, 'Muriel', 'Sult Rettig', 4158),
 (82, 'George', 'Nejaime Perez', 8499),
 (102, 'Ray', 'Outen Spruiell', 5576),
 (121, 'Tina', 'Henning Davis', 8607),
 (67, 'Jennifer', 'Balboa Ramos', 433),
 (108, 'Andria', 'Hanson Camacho', 3800),
 (111, 'Eliza', 'Starr Molina', 7516),
 (131, 'Aaron', 'Gray Buford', 8417),
 (156, 'Sean', 'Horst Townsend', 1285),
 (163, 'Kevin', 'Barajas Dipietro', 4422),
 (163, 'Kevin', 'Barajas Dipietro', 2178),
 (173, 'Sandra', 'Horn Calhoun', 4087),
 (174, 'Juli', 'Sarmiento Pujols', 3083),
 (202, 'Emma', 'Weathers Halstead', 5264),
 (203, 'Rickey', 'Edwards Alexander', 4549)]

### Consulta 4 (6 pts)
**Nombre de la Facción y la cantidad de trabajadores activos hasta fecha X.**

Ejemplo:

| Facción     | Cantidad de trabajadores |
|----------------|--------|
| FEDRA | 1273 |
| Luciérnagas | 121 |
| Carroñeros  | 72 |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [28]:
%%sql

SELECT facciones.nombre AS faccion, COUNT (persona_tiene_trabajo.id_persona) AS cantidad_de_trabajadores
FROM
    facciones

RIGHT JOIN
     persona_pertenece_faccion
ON
    facciones.id=persona_pertenece_faccion.id_faccion

LEFT JOIN
    persona_tiene_trabajo
ON
    persona_pertenece_faccion.id_persona=persona_tiene_trabajo.id_persona

WHERE
    persona_tiene_trabajo.estado=true

GROUP BY
    facciones.nombre;

 * postgresql://postgres:***@localhost:5432/tareabase
6 rows affected.


[('JavaScriptia', 5),
 ('Luciérnagas', 8),
 ('Pythonia', 4),
 ('MaCdonia', 7),
 ('FEDRA', 8),
 ('Carroñeros', 12)]

### Consulta 5 (6 pts)
**Nombre de la Facción y su cantidad de personas infectadas.**

Ejemplo:

|  Facción  | Personas infectadas | 
|--------|-------|
| FEDRA | 251 |
| Luciérnagas | 17 |
| Carroñeros  | 12 |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [173]:
%%sql


SELECT facciones.nombre AS faccion, COUNT(personas.id) AS personas_infectadas
FROM 
    facciones
LEFT JOIN
    persona_pertenece_faccion 
ON 
    facciones.id = persona_pertenece_faccion.id_faccion

LEFT JOIN 
    personas 
ON 
    persona_pertenece_faccion.id_persona = personas.id

WHERE personas.infectado = true
GROUP BY facciones.nombre;


 * postgresql://postgres:***@localhost:5432/tarea-bd
6 rows affected.


[('JavaScriptia', 26),
 ('Luciérnagas', 20),
 ('Pythonia', 20),
 ('MaCdonia', 21),
 ('FEDRA', 29),
 ('Carroñeros', 28)]

### Consulta 6 (6 pts)
**Nombre de la Facción y su trabajo más realizado.**

Ejemplo:

|  Facción  | Trabajo | 
|--------|-------|
| FEDRA | Limpiador de cloacas |
| Luciérnagas | Médico |
| Carroñeros  | Explorador |
| ...  | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [43]:
%%sql


SELECT
    facciones.nombre AS Facción  ,trabajos.descripcion ,
    row_number() over (partition by facciones.nombre order by COUNT(persona_tiene_trabajo.id_trabajo) DESC) rn
FROM 
    facciones

LEFT JOIN
    persona_pertenece_faccion
ON 
    facciones.id = persona_pertenece_faccion.id_faccion

LEFT JOIN
    personas
ON
    persona_pertenece_faccion.id_persona = personas.id

LEFT JOIN
    persona_tiene_trabajo
ON
    personas.id= persona_tiene_trabajo.id_persona
INNER JOIN
    trabajos
ON
    persona_tiene_trabajo.id_trabajo= trabajos.id

WHERE persona_tiene_trabajo.estado=true

GROUP BY facciones.id ,  trabajos.descripcion

ORDER BY rn ASC

LIMIT 6;


 * postgresql://postgres:***@localhost:5432/tareabase
6 rows affected.


[('Luciérnagas', 'Encargado de construir', 1),
 ('MaCdonia', 'encargado de curar a los enfermos', 1),
 ('Carroñeros', 'Encargado de explorar la ciudad', 1),
 ('JavaScriptia', 'Encargado de proteger la ciudad', 1),
 ('FEDRA', 'Encargado de explorar la ciudad', 1),
 ('Pythonia', 'Encargado de limpiar las calles', 1)]

### Consulta 7 (8 pts)
**Nombre de la Facción y su Saldo total** (Es decir, la suma total de los saldos de cada persona perteneciente a la facción.)

Ejemplo:

|Facción | Saldo |
|--------|-|
| FEDRA | 523000 |
| Luciérnagas | 130023 |
| Carroñeros  | 17432 |
| ...  | ...|

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

SELECT
    facciones.nombre AS Facción, SUM(cuentas_bancarias.saldo) AS Saldo

FROM 
    facciones

LEFT JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion

LEFT JOIN personas ON persona_pertenece_faccion.id_persona = personas.id

LEFT JOIN cuentas_bancarias ON personas.id = cuentas_bancarias.id_persona

GROUP BY facciones.nombre;

### Consulta 8 (6 pts)
**Nombre del Banco, cantidad de usuarios pertenecientes al banco y saldo total del banco.**

###### Es decir, todos los bancos que ocupan los habitantes de cada faccion

Ejemplo:

|Banco | Cantidad|Saldo|
|----|---|---|
|FedraBank S.A |321|20000|
|Banco Pedrito Pascal SpA  |21|200| 
| Banco De Chile|213|56000|
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

SELECT
    cuentas_bancarias.banco AS Banco , COUNT(DISTINCT personas.id) AS Cantidad, SUM(cuentas_bancarias.saldo) AS Saldo
FROM 
    cuentas_bancarias
INNER JOIN personas ON cuentas_bancarias.id_persona = personas.id

GROUP BY cuentas_bancarias.banco;

### Consulta 9 (8 pts)
**Nombre de la Facción y el Banco más utilizado por sus habitantes**(Si hay empate, que se muestre solo uno.)

Ejemplo:
   
|Facción | Banco |
| - | - |
| FEDRA | FedraBank S.A |
| Luciérnagas | Banco Pedrito Pascal SpA |
| ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [365]:
%%sql

SELECT facciones.nombre , cuentas_bancarias.tipo_de_cuenta , COUNT(cuentas_bancarias.tipo_de_cuenta) AS contador,
row_number() over (partition by facciones.nombre order by COUNT(cuentas_bancarias.tipo_de_cuenta) DESC) rn
FROM
    facciones

LEFT JOIN
    persona_pertenece_faccion
ON
    facciones.id = persona_pertenece_faccion.id_faccion

LEFT JOIN
    personas
ON
    persona_pertenece_faccion.id_persona = personas.id

INNER JOIN
    cuentas_bancarias
ON
    personas.id=cuentas_bancarias.id_persona
GROUP BY facciones.nombre , cuentas_bancarias.tipo_de_cuenta 

ORDER BY rn
Limit 6

 * postgresql://postgres:***@localhost:5432/tarea-bd
6 rows affected.


[('MaCdonia', 'Cuenta Vista', 1, 1),
 ('JavaScriptia', 'Cuenta de Credito', 3, 1),
 ('Carroñeros', 'Cuenta Corriente', 3, 1),
 ('Luciérnagas', 'Cuenta de Inversion', 2, 1),
 ('FEDRA', 'Cuenta de Inversion', 4, 1),
 ('Pythonia', 'Cuenta de Inversion', 2, 1)]

### Consulta 10 (6 pts)
**ID, nombre y segundo apellido de las personas traidoras**(Es decir, si la persona pertenece a dos o más facciones.)

Ejemplo:
   
|ID | Nombre | Apellido | 
|-|-------|-|
| 1 | Pedro | Pascal |
| 2 | Bella | Ramsey |
| ... | ... | ... | 

**Solución:** _Escriba aquí la descripción de su solución._

In [348]:
%%sql

SELECT
    personas.id,facciones.id AS FAC,personas.nombre,row_number() over (partition by personas.id order by personas.id ASC) rn
FROM 
    facciones

LEFT JOIN
    persona_pertenece_faccion
ON
    facciones.id=persona_pertenece_faccion.id_faccion

LEFT JOIN
    personas
ON
    persona_pertenece_faccion.id_persona=personas.id

GROUP BY personas.id,facciones.id , personas.nombre ,personas.id
ORDER BY rn DESC


 * postgresql://postgres:***@localhost:5432/tarea-bd
295 rows affected.


[(24, 6, 'Estela', 3),
 (596, 6, 'Robert', 3),
 (427, 4, 'Alejandro', 3),
 (201, 6, 'Everett', 3),
 (659, 2, 'Charles', 2),
 (169, 6, 'Anna', 2),
 (897, 3, 'James', 2),
 (249, 5, 'Michele', 2),
 (448, 5, 'John', 2),
 (339, 6, 'Florentino', 2),
 (905, 4, 'Charles', 2),
 (201, 3, 'Everett', 2),
 (251, 5, 'Naomi', 2),
 (691, 4, 'Carlton', 2),
 (793, 5, 'Ray', 2),
 (596, 3, 'Robert', 2),
 (522, 4, 'Jose', 2),
 (185, 6, 'Charlotte', 2),
 (608, 2, 'Shonda', 2),
 (708, 6, 'Randy', 2),
 (21, 5, 'Jorge', 2),
 (425, 5, 'Caitlyn', 2),
 (24, 4, 'Estela', 2),
 (722, 5, 'Wan', 2),
 (427, 3, 'Alejandro', 2),
 (625, 6, 'Charles', 2),
 (735, 2, 'Norman', 2),
 (549, 6, 'James', 2),
 (366, 5, 'Diane', 2),
 (183, 4, 'Dale', 2),
 (233, 6, 'Carl', 2),
 (236, 6, 'John', 2),
 (766, 4, 'Ana', 2),
 (375, 4, 'Gary', 2),
 (121, 2, 'Tawny', 1),
 (123, 3, 'Lester', 1),
 (127, 3, 'Alejandro', 1),
 (128, 1, 'Tracy', 1),
 (130, 5, 'Malcom', 1),
 (134, 4, 'Douglas', 1),
 (136, 3, 'William', 1),
 (141, 3, 'John', 1),
 (153, 6, 'Jason', 1),
 (158, 3, 'Paulette', 1),
 (161, 1, 'Shannon', 1),
 (163, 3, 'Jeannie', 1),
 (164, 6, 'Robert', 1),
 (169, 5, 'Anna', 1),
 (170, 2, 'Margaret', 1),
 (180, 4, 'David', 1),
 (183, 1, 'Dale', 1),
 (185, 4, 'Charlotte', 1),
 (186, 6, 'Kenneth', 1),
 (191, 6, 'Jason', 1),
 (193, 2, 'Christina', 1),
 (195, 1, 'Glenn', 1),
 (201, 1, 'Everett', 1),
 (202, 3, 'Anita', 1),
 (207, 3, 'Michael', 1),
 (209, 3, 'Paul', 1),
 (211, 6, 'Jeremy', 1),
 (214, 2, 'Travis', 1),
 (215, 6, 'Katie', 1),
 (218, 5, 'Scott', 1),
 (225, 3, 'James', 1),
 (226, 5, 'David', 1),
 (230, 6, 'Jerry', 1),
 (231, 2, 'Ann', 1),
 (233, 2, 'Carl', 1),
 (236, 1, 'John', 1),
 (249, 2, 'Michele', 1),
 (251, 4, 'Naomi', 1),
 (255, 5, 'Rodney', 1),
 (257, 1, 'Stacey', 1),
 (259, 1, 'Kathy', 1),
 (265, 1, 'Hans', 1),
 (267, 1, 'Juan', 1),
 (274, 4, 'Ruth', 1),
 (275, 3, 'Wanda', 1),
 (282, 3, 'Marguerite', 1),
 (285, 4, 'Janine', 1),
 (294, 1, 'Ruth', 1),
 (295, 5, 'Helen', 1),
 (300, 3, 'Kelly', 1),
 (304, 5, 'Monique', 1),
 (306, 1, 'Samuel', 1),
 (309, 1, 'Erika', 1),
 (311, 6, 'Nancy', 1),
 (312, 3, 'Yolanda', 1),
 (326, 1, 'Sean', 1),
 (328, 1, 'Lissette', 1),
 (331, 1, 'Joseph', 1),
 (335, 6, 'Jean', 1),
 (339, 4, 'Florentino', 1),
 (340, 4, 'Damon', 1),
 (342, 2, 'Effie', 1),
 (343, 5, 'Brandon', 1),
 (355, 6, 'Julie', 1),
 (359, 2, 'Yvonne', 1),
 (362, 5, 'Jerry', 1),
 (366, 4, 'Diane', 1),
 (367, 2, 'Margaret', 1),
 (374, 3, 'Gloria', 1),
 (375, 2, 'Gary', 1),
 (379, 2, 'Forrest', 1),
 (384, 3, 'Roberta', 1),
 (390, 4, 'Myra', 1),
 (395, 2, 'Andrew', 1),
 (404, 1, 'Richard', 1),
 (408, 1, 'Kayla', 1),
 (419, 3, 'Nicole', 1),
 (425, 3, 'Caitlyn', 1),
 (427, 2, 'Alejandro', 1),
 (428, 1, 'Sherry', 1),
 (429, 5, 'Albert', 1),
 (430, 3, 'Jean', 1),
 (434, 2, 'Crystal', 1),
 (436, 4, 'Bruce', 1),
 (437, 6, 'Carl', 1),
 (447, 3, 'Paul', 1),
 (448, 4, 'John', 1),
 (449, 1, 'Luther', 1),
 (456, 5, 'Paul', 1),
 (462, 2, 'Linda', 1),
 (463, 4, 'Gary', 1),
 (466, 3, 'Janet', 1),
 (468, 3, 'Patricia', 1),
 (470, 2, 'Sheila', 1),
 (474, 1, 'Donna', 1),
 (476, 1, 'Margot', 1),
 (483, 6, 'Wendell', 1),
 (484, 6, 'William', 1),
 (488, 3, 'Michelle', 1),
 (490, 3, 'Bradley', 1),
 (491, 2, 'Bernice', 1),
 (497, 4, 'Hilda', 1),
 (498, 4, 'Virginia', 1),
 (499, 5, 'Sara', 1),
 (500, 1, 'Kim', 1),
 (501, 5, 'Christian', 1),
 (506, 1, 'Jo', 1),
 (509, 6, 'Dennis', 1),
 (513, 6, 'Luis', 1),
 (522, 1, 'Jose', 1),
 (525, 6, 'Frederick', 1),
 (527, 2, 'Donald', 1),
 (533, 5, 'Thomas', 1),
 (534, 5, 'Huey', 1),
 (536, 1, 'Mary', 1),
 (541, 1, 'Mel', 1),
 (549, 5, 'James', 1),
 (550, 1, 'Travis', 1),
 (553, 4, 'William', 1),
 (554, 3, 'Larry', 1),
 (562, 6, 'Emily', 1),
 (563, 6, 'Francisco', 1),
 (568, 6, 'Stan', 1),
 (575, 5, 'Donald', 1),
 (577, 5, 'Misty', 1),
 (579, 3, 'Aaron', 1),
 (592, 3, 'Sarah', 1),
 (595, 2, 'Anthony', 1),
 (596, 2, 'Robert', 1),
 (607, 3, 'Doug', 1),
 (608, 1, 'Shonda', 1),
 (617, 2, 'Ramon', 1),
 (621, 2, 'Sandy', 1),
 (624, 6, 'Elmer', 1),
 (625, 2, 'Charles',

### Consulta BONUS (6 pts.)

**Nombre y Apellidos de las personas casadas**(solo se muestra una única vez la pareja)

Ejemplo:
   
|Nombre persona 1 | Apellidos persona 1 | Nombre persona 2 | Apellidos persona 2 |
|-|-------|-|-|
| Felipe | Martinez | Raul | Cuello |
| Rodrigo | Munita | Juan | Mira |
| Nicole | Rojel | Benjamin | Campos |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [ ]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *